src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -0,0 +1,136 @@
+ *
+ *  1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Heiko Strathmann, Viktor Gal
+ */
+
+#include <shogun/mathematics/linalg/LinalgBackendBase.h>
+#include <shogun/mathematics/linalg/SGLinalg.h>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+/** Return the corresponding backend of SGVector
+ * @param vec SGVector
+ * @return LinalgBackendBase pointer
+ */
+template <typename Type>
+LinalgBackendBase* get_backend(const SGVector<Type>& vec)
+{
+	if (vec.on_gpu())
+		return sg_linalg->get_gpu_backend();
+	else
+		return sg_linalg->get_cpu_backend();
+}
+
+/** Return the corresponding backend of SGVector
+ * Raise error if the two SGVectors are not on the same backend.
+ * @param a the fisrt SGVector
+ * @param b the second SGVector
+ * @return LinalgBackendBase pointer
+ */
+template <typename Type>
+LinalgBackendBase* get_backend(const SGVector<Type>& a, const SGVector<Type>& b)
+{
+	if (a.on_gpu() && b.on_gpu())
+		return sg_linalg->get_gpu_backend();
+	else if (a.on_gpu() || b.on_gpu())
+		SG_SERROR("Cannot operate with first vector gpu (%d) and second vector gpu (%d).\n",
+					a.on_gpu(), b.on_gpu());
+					
+	return sg_linalg->get_cpu_backend();
+}
+
+/**
+ * Vector dot-product that works with generic vectors.
+ *
+ * @param a first vector
+ * @param b second vector
+ * @return the dot product of \f$\mathbf{a}\f$ and \f$\mathbf{b}\f$, represented
+ * as \f$\sum_i a_i b_i\f$
+ */
+template <typename Type>
+Type dot(const SGVector<Type>& a, const SGVector<Type>& b)
+{
+	return get_backend(a, b)->dot(a, b);
```

Comment:
I was wrong, actually this is good, as it can be re-used in other places.
Nice!

What you should do is do a null check, and throw an error in that case
(this is when data was transfered to GPU and then the gpu backend was set to null)